planner/core/rule_partition_processor.go

```diff
@@ -224,13 +224,80 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
 	return tableDual, nil
 }
 
-func (s *partitionProcessor) pruneListPartition(tbl table.Table, partitionNames []model.CIStr) []int {
-	if len(partitionNames) > 0 {
-		pi := tbl.Meta().Partition
+func (s *partitionProcessor) findUsedListPartitions(ctx sessionctx.Context, tbl table.Table, partitionNames []model.CIStr,
+	conds []expression.Expression) ([]int, error) {
+	pi := tbl.Meta().Partition
+	partExpr, err := tbl.(partitionTable).PartitionExpr()
+	if err != nil {
+		return nil, err
+	}
+	pruneList := partExpr.ForListPruning
+	cols := make([]*expression.Column, 0, len(pruneList.ExprCols))
+	colLen := make([]int, 0, len(pruneList.ExprCols))
+	for _, c := range pruneList.ExprCols {
+		cols = append(cols, c.Clone().(*expression.Column))
+		colLen = append(colLen, types.UnspecifiedLength)
+	}
+
+	datchedResult, err := ranger.DetachCondAndBuildRangeForPartition(ctx, conds, cols, colLen)
+	if err != nil {
+		return nil, err
+	}
+
+	ranges := datchedResult.Ranges
+	used := make([]int, 0, len(ranges))
+	for _, r := range ranges {
+		if r.IsPointNullable(ctx.GetSessionVars().StmtCtx) {
+			if !r.HighVal[0].IsNull() {
+				if len(r.HighVal) != len(cols) {
+					used = []int{-1}
+					break
+				}
+			}
+			found := int64(-1)
+			for j, expr := range pruneList.Exprs {
+				ret, _, err := expr.EvalInt(ctx, chunk.MutRowFromDatums(r.HighVal).ToRow())
```

Comment:
Maybe I can't count on this function to be high-performance.
There are so many allocations if you take a close look.

Reply:
Great catch, done.